In [ ]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

# Use the kagglehub client library to attach Kaggle resources like competitions, datasets, and models to your session
# Learn more about kagglehub: https://github.com/Kaggle/kagglehub/blob/main/README.md

import kagglehub
# kagglehub.dataset_download('<owner>/<dataset-slug>')

In [1]:
import numpy as np
import pandas as pd

In [2]:
train=pd.read_csv('/kaggle/input/competitions/smart-mcq-solver-challenge/train.csv')
test=pd.read_csv('/kaggle/input/competitions/smart-mcq-solver-challenge/test.csv')
sample_submission=pd.read_csv('/kaggle/input/competitions/smart-mcq-solver-challenge/sample_submission.csv')

In [4]:
train.shape,test.shape

((2000, 8), (500, 7))

In [34]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 8 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      2000 non-null   int64 
 1   prompt  2000 non-null   object
 2   A       2000 non-null   object
 3   B       2000 non-null   object
 4   C       2000 non-null   object
 5   D       2000 non-null   object
 6   E       2000 non-null   object
 7   answer  2000 non-null   object
dtypes: int64(1), object(7)
memory usage: 125.1+ KB


In [ ]:
test.info()

In [3]:
train.head(3)

,id,prompt,A,B,C,D,E,answer
0,1,Pick the best possible answer: What is Martin ...,Martin Heidegger believes that humans exist wi...,Martin Heidegger believes that humans do not e...,Martin Heidegger does not believe in the exist...,Martin Heidegger believes that the relationshi...,Martin Heidegger believes that time is an illu...,B
1,2,What is accelerator-based light-ion fusion?,Accelerator-based light-ion fusion is a techni...,Accelerator-based light-ion fusion is a techni...,Accelerator-based light-ion fusion is a techni...,Accelerator-based light-ion fusion is a techni...,Accelerator-based light-ion fusion is a techni...,A
2,3,Determine the correct option: What is the term...,Blueshifting,Redshifting,Reddening,Whitening,Yellowing,C


===================================================================

# MODEL_NAME = "BAAI/bge-large-en-v1.5" (Pretrained model)

In [ ]:
# # Imports

# import os
# import gc
# import random
# import warnings

# import numpy as np
# import pandas as pd

# from tqdm.auto import tqdm

# from sentence_transformers import SentenceTransformer

# from sklearn.linear_model import LogisticRegression

# from sklearn.model_selection import train_test_split

# from sklearn.metrics import (
#     accuracy_score,
#     precision_score,
#     recall_score,
#     f1_score,
#     roc_auc_score,
#     confusion_matrix,
#     classification_report
# )

# warnings.filterwarnings("ignore")

In [ ]:
# import wandb
# from kaggle_secrets import UserSecretsClient

# user_secrets = UserSecretsClient()

# wandb.login(
#     key=user_secrets.get_secret("WANDB_API_KEY")
# )

# wandb.init(
#     project="mcq-challenge",
#     name="bge-large-logreg"
# )

#Model name and seed

In [ ]:
# MODEL_NAME = "BAAI/bge-large-en-v1.5"

# SEED = 42

# random.seed(SEED)
# np.random.seed(SEED)

In [ ]:
# # Answer Options

# OPTIONS = ["A", "B", "C", "D", "E"]

In [ ]:
# # Build Binary Training Dataset

# def build_train(df):

#     texts = []
#     labels = []
#     qids = []
#     options = []

#     for _, row in tqdm(df.iterrows(), total=len(df)):

#         prompt = str(row["prompt"])

#         answer = row["answer"]

#         for op in OPTIONS:

#             candidate = str(row[op])

#             text = f"""
# Question:
# {prompt}

# Candidate Answer:
# {candidate}
# """

#             texts.append(text)

#             labels.append(int(op == answer))

#             qids.append(row["id"])

#             options.append(op)

#     return texts, np.array(labels), qids, options

In [ ]:
# train_texts, train_labels, train_qids, train_options = build_train(train)

# print("Training samples:", len(train_texts))
# print("Positive labels :", train_labels.sum())
# print("Negative labels :", len(train_labels) - train_labels.sum())

In [ ]:
# # Build Test Dataset
# # --------------------------------------

# def build_test(df):

#     texts = []
#     qids = []
#     options = []

#     for _, row in tqdm(df.iterrows(), total=len(df)):

#         prompt = str(row["prompt"])

#         for op in OPTIONS:

#             candidate = str(row[op])

#             text = f"""
# Question:
# {prompt}

# Candidate Answer:
# {candidate}
# """

#             texts.append(text)

#             qids.append(row["id"])

#             options.append(op)

#     return texts, qids, options

In [ ]:
# test_texts, test_qids, test_options = build_test(test)

# print("Test samples:", len(test_texts))

In [ ]:
# print(pd.Series(train_labels).value_counts())

In [ ]:
# sample_idx = 0

# print("Question ID :", train_qids[sample_idx])
# print("Option      :", train_options[sample_idx])
# print("Label       :", train_labels[sample_idx])
# print()
# print(train_texts[sample_idx])

In [ ]:
# # Sentence Embeddings
# #-----------------------------------------

# from sentence_transformers import SentenceTransformer
# import torch

In [ ]:
# device = "cuda" if torch.cuda.is_available() else "cpu"

# print("Device:", device)

In [ ]:
# MODEL_NAME = "BAAI/bge-large-en-v1.5"

# embedder = SentenceTransformer(
#     MODEL_NAME,
#     device=device
# )

# print("Embedding model loaded.")

#Train embeddings

#Converts each text into a dense embedding vector.

In [ ]:
# X_train = embedder.encode(
#     train_texts,
#     batch_size=32,
#     show_progress_bar=True,
#     convert_to_numpy=True,
#     normalize_embeddings=True
# )

# print("Train embeddings shape:", X_train.shape)

#test embeddings

In [ ]:
# X_test = embedder.encode(
#     test_texts,
#     batch_size=32,
#     show_progress_bar=True,
#     convert_to_numpy=True,
#     normalize_embeddings=True
# )

# print("Test embeddings shape:", X_test.shape)

In [ ]:
# np.save("X_train.npy", X_train)
# np.save("X_test.npy", X_test)

# print("Embeddings saved.")

In [ ]:
# X_train = np.load("X_train.npy")
# X_test = np.load("X_test.npy")

In [ ]:
# import gc

# gc.collect()

# if torch.cuda.is_available():
#     torch.cuda.empty_cache()

#train_test_split

In [ ]:
# from sklearn.model_selection import train_test_split

# train_ids, valid_ids = train_test_split(
#     train["id"],
#     test_size=0.2,
#     random_state=42,
#     shuffle=True
# )

# train_ids = set(train_ids)
# valid_ids = set(valid_ids)

In [ ]:
# train_mask = np.array([qid in train_ids for qid in train_qids])
# valid_mask = np.array([qid in valid_ids for qid in train_qids])

# X_tr = X_train[train_mask]
# X_val = X_train[valid_mask]

# y_tr = train_labels[train_mask]
# y_val = train_labels[valid_mask]

# qid_val = np.array(train_qids)[valid_mask]
# option_val = np.array(train_options)[valid_mask]

# print(X_tr.shape)
# print(X_val.shape)

#Logistic Regression classifier Learns to classify the embeddings into target classes.

In [ ]:
# from sklearn.linear_model import LogisticRegression

# clf = LogisticRegression(

#     C=4,

#     max_iter=5000,

#     class_weight="balanced",

#     solver="lbfgs",

#     random_state=42,

#     n_jobs=-1

# )

# clf.fit(X_tr, y_tr)

In [ ]:
# val_prob = clf.predict_proba(X_val)[:,1]

# val_pred = (val_prob >= 0.5).astype(int)

In [ ]:
# from sklearn.metrics import accuracy_score

# acc = accuracy_score(y_val, val_pred)

# print("Accuracy:", acc)

In [ ]:
# from sklearn.metrics import f1_score

# f1 = f1_score(y_val, val_pred)

# print("F1:", f1)

In [ ]:
# # Validation Predictions
# # --------------------------------------------------

# valid_df = pd.DataFrame({
#     "id": qid_val,
#     "option": option_val,
#     "label": y_val,
#     "score": val_prob
# })

# valid_df.head()

In [ ]:
# valid_df = valid_df.sort_values(
#     ["id", "score"],
#     ascending=[True, False]
# )

# valid_df.head(10)

#Map3 function

In [ ]:
# def map3(df):

#     total = 0.0

#     groups = df.groupby("id")

#     for _, g in groups:

#         pred = g["option"].tolist()[:3]

#         gt = g.loc[g["label"] == 1, "option"].values[0]

#         if pred[0] == gt:
#             total += 1.0

#         elif len(pred) > 1 and pred[1] == gt:
#             total += 0.5

#         elif len(pred) > 2 and pred[2] == gt:
#             total += 1.0 / 3.0

#     return total / len(groups)

In [ ]:
# map3_score = map3(valid_df)

# print(f"MAP@3 : {map3_score:.5f}")

In [ ]:
# top3 = (
#     valid_df
#     .groupby("id")["option"]
#     .apply(lambda x: " ".join(x.head(3)))
#     .reset_index()
# )

# top3.head()

In [ ]:
# wandb.log({
#     "Accuracy": acc,
#     "F1": f1,
#     "MAP@3": map3_score
# })

In [ ]:
# # Train Final Logistic Regression Model
# # ----------------------------------------------

# from sklearn.linear_model import LogisticRegression

# final_model = LogisticRegression(
#     C=4,
#     max_iter=5000,
#     class_weight="balanced",
#     solver="lbfgs",
#     random_state=42,
#     n_jobs=-1
# )

# final_model.fit(X_train, train_labels)

# print("Training Complete!")

In [ ]:
# # Predict Test
# # -----------------------------------------------

# test_prob = final_model.predict_proba(X_test)[:,1]

# print(test_prob.shape)

In [ ]:
# pred_df = pd.DataFrame({

#     "id": test_qids,

#     "option": test_options,

#     "score": test_prob

# })

# pred_df.head()

In [ ]:
# pred_df = pred_df.sort_values(

#     ["id","score"],

#     ascending=[True,False]

# )

In [ ]:
# submission = (

#     pred_df.groupby("id")["option"]

#     .apply(lambda x: " ".join(x.head(3)))

#     .reset_index()

# )

# submission.columns=["id","Prediction"]

# submission.head()

In [ ]:
# submission.to_csv(

#     "submission.csv",

#     index=False

# )

# print("submission.csv saved.")

#In the above pretrained model serves as a feature extractor, and Logistic Regression is the downstream classifier.

================================================

# Model-2:
**LightGBM MCQ Ranking Baseline**

In [8]:
import wandb

wandb.login(key="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")

wandb: WARNING If you're specifying your api key in code, ensure this code is not shared publicly.
wandb: WARNING Consider setting the WANDB_API_KEY environment variable, or running `wandb login` from the command line.
wandb: WARNING [wandb.login()] Changing session credentials to explicit value for https://api.wandb.ai.
wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc


True

In [3]:
import numpy as np
import pandas as pd
import lightgbm as lgb
from scipy.sparse import hstack
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import accuracy_score, f1_score
import wandb

#Expand each question into question–choice pairs

In [4]:
labels = ["A","B","C","D","E"]
def expand(df, train_mode=True):
    rows=[]
    for _,r in df.iterrows():
        for c in labels:
            rows.append({
                "id":r["id"],
                "choice":c,
                "text":f"Question: {r['prompt']} Answer: {r[c]}",
                "y":1 if train_mode and r["answer"]==c else 0
            })
    return pd.DataFrame(rows)


#TF-IDF Feature Extraction and Cross-Validation Setup

In [5]:
tr = expand(train,True)
te = expand(test,False)
tfidf = TfidfVectorizer(
    max_features=80000,
    ngram_range=(1,2),
    stop_words="english",
    sublinear_tf=True
)

X_all = tfidf.fit_transform(pd.concat([tr.text,te.text]))
Xtr = X_all[:len(tr)]
Xte = X_all[len(tr):]

y = tr.y.values
groups = tr.groupby("id")["y"].max().values

skf = StratifiedKFold(n_splits=5,shuffle=True,random_state=42)

oof=np.zeros(len(tr))
pred=np.zeros(len(te))

unique_ids=train.id.values

#light gbm 

In [6]:
for fold,(tr_idx,val_idx) in enumerate(skf.split(unique_ids,groups)):
    wandb.init(project="mcq-lightgbm",name=f"fold-{fold}",reinit=True)

    train_ids=set(unique_ids[tr_idx])
    val_ids=set(unique_ids[val_idx])

    tr_rows=tr.id.isin(train_ids).values
    va_rows=tr.id.isin(val_ids).values

    model=lgb.LGBMClassifier(
        objective="binary",
        n_estimators=700,
        learning_rate=0.03,
        num_leaves=63,
        feature_fraction=0.8,
        bagging_fraction=0.8,
        bagging_freq=5,
        random_state=42
    )

    model.fit(Xtr[tr_rows],y[tr_rows])

    p=model.predict_proba(Xtr[va_rows])[:,1]
    oof[va_rows]=p

    acc=accuracy_score(y[va_rows],p>0.5)
    f1=f1_score(y[va_rows],p>0.5)

    wandb.log({"accuracy":acc,"f1":f1})

    pred+=model.predict_proba(Xte)[:,1]/5

te["score"]=pred




/usr/local/lib/python3.12/dist-packages/notebook/notebookapp.py:191: SyntaxWarning: invalid escape sequence '\/'
  | |_| | '_ \/ _` / _` |  _/ -_)
wandb: (1) Create a W&B account
wandb: (2) Use an existing W&B account
wandb: (3) Don't visualize my results
wandb: Enter your choice:

  2


wandb: You chose 'Use an existing W&B account'
wandb: Logging into https://api.wandb.ai. (Learn how to deploy a W&B server locally: https://wandb.me/wandb-server)
wandb: Create a new API key at: https://wandb.ai/authorize?ref=models
wandb: Store your API key securely and do not share it.
wandb: Paste your API key and hit enter:

  ········


wandb: No netrc file found, creating one.
wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc
wandb: Currently logged in as: 24f2006126 (24f2006126-iitm) to https://api.wandb.ai. Use `wandb login --relogin` to force relogin
wandb: WARNING Using a boolean value for 'reinit' is deprecated. Use 'return_previous' or 'finish_previous' instead.


[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] feature_fraction is set=0.8, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8
[LightGBM] [Warning] bagging_fraction is set=0.8, subsample=1.0 will be ignored. Current value: bagging_fraction=0.8
[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] feature_fraction is set=0.8, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8
[LightGBM] [Warning] bagging_fraction is set=0.8, subsample=1.0 will be ignored. Current value: bagging_fraction=0.8
[LightGBM] [Info] Number of positive: 1600, number of negative: 6400
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.049601 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 85705
[LightGBM] [Info] Number of data poi

/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


[LightGBM] [Warning] feature_fraction is set=0.8, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8
[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] bagging_fraction is set=0.8, subsample=1.0 will be ignored. Current value: bagging_fraction=0.8
[LightGBM] [Warning] feature_fraction is set=0.8, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8
[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] bagging_fraction is set=0.8, subsample=1.0 will be ignored. Current value: bagging_fraction=0.8


accuracy,▁
f1,▁
accuracy,0.967
f1,0.91582


[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] feature_fraction is set=0.8, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8
[LightGBM] [Warning] bagging_fraction is set=0.8, subsample=1.0 will be ignored. Current value: bagging_fraction=0.8
[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] feature_fraction is set=0.8, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8
[LightGBM] [Warning] bagging_fraction is set=0.8, subsample=1.0 will be ignored. Current value: bagging_fraction=0.8
[LightGBM] [Info] Number of positive: 1600, number of negative: 6400
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.051202 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 85159
[LightGBM] [Info] Number of data poi

/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


[LightGBM] [Warning] feature_fraction is set=0.8, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8
[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] bagging_fraction is set=0.8, subsample=1.0 will be ignored. Current value: bagging_fraction=0.8
[LightGBM] [Warning] feature_fraction is set=0.8, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8
[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] bagging_fraction is set=0.8, subsample=1.0 will be ignored. Current value: bagging_fraction=0.8


accuracy,▁
f1,▁
accuracy,0.9555
f1,0.87957


[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] feature_fraction is set=0.8, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8
[LightGBM] [Warning] bagging_fraction is set=0.8, subsample=1.0 will be ignored. Current value: bagging_fraction=0.8
[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] feature_fraction is set=0.8, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8
[LightGBM] [Warning] bagging_fraction is set=0.8, subsample=1.0 will be ignored. Current value: bagging_fraction=0.8
[LightGBM] [Info] Number of positive: 1600, number of negative: 6400
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.059519 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 86707
[LightGBM] [Info] Number of data poi

/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


[LightGBM] [Warning] feature_fraction is set=0.8, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8
[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] bagging_fraction is set=0.8, subsample=1.0 will be ignored. Current value: bagging_fraction=0.8
[LightGBM] [Warning] feature_fraction is set=0.8, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8
[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] bagging_fraction is set=0.8, subsample=1.0 will be ignored. Current value: bagging_fraction=0.8


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


accuracy,▁
f1,▁
accuracy,0.9545
f1,0.87818


[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] feature_fraction is set=0.8, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8
[LightGBM] [Warning] bagging_fraction is set=0.8, subsample=1.0 will be ignored. Current value: bagging_fraction=0.8
[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] feature_fraction is set=0.8, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8
[LightGBM] [Warning] bagging_fraction is set=0.8, subsample=1.0 will be ignored. Current value: bagging_fraction=0.8
[LightGBM] [Info] Number of positive: 1600, number of negative: 6400
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.050355 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 85962
[LightGBM] [Info] Number of data poi

/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


[LightGBM] [Warning] feature_fraction is set=0.8, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8
[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] bagging_fraction is set=0.8, subsample=1.0 will be ignored. Current value: bagging_fraction=0.8
[LightGBM] [Warning] feature_fraction is set=0.8, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8
[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] bagging_fraction is set=0.8, subsample=1.0 will be ignored. Current value: bagging_fraction=0.8


accuracy,▁
f1,▁
accuracy,0.959
f1,0.89295


[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] feature_fraction is set=0.8, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8
[LightGBM] [Warning] bagging_fraction is set=0.8, subsample=1.0 will be ignored. Current value: bagging_fraction=0.8
[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] feature_fraction is set=0.8, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8
[LightGBM] [Warning] bagging_fraction is set=0.8, subsample=1.0 will be ignored. Current value: bagging_fraction=0.8
[LightGBM] [Info] Number of positive: 1600, number of negative: 6400
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.097487 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 85310
[LightGBM] [Info] Number of data poi

/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


[LightGBM] [Warning] feature_fraction is set=0.8, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8
[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] bagging_fraction is set=0.8, subsample=1.0 will be ignored. Current value: bagging_fraction=0.8
[LightGBM] [Warning] feature_fraction is set=0.8, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8
[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] bagging_fraction is set=0.8, subsample=1.0 will be ignored. Current value: bagging_fraction=0.8


accuracy,▁
f1,▁
accuracy,0.958
f1,0.88889


#submission file

In [7]:
sub=[]
for qid,g in te.groupby("id"):
    top=g.sort_values("score",ascending=False).head(3)["choice"].tolist()
    sub.append([qid," ".join(top)])

submission=pd.DataFrame(sub,columns=["ID","Prediction"])
submission.to_csv("submission.csv",index=False)

print(submission.head())
print("Saved submission.csv")

   ID Prediction
0   1      A E B
1   2      B E D
2   3      B E D
3   4      E C D
4   5      C B D
Saved submission.csv


Kaggle score =0.738

# Model-1:

**Attention-based BiLSTM Ranker trained from scratch using BCE loss.**

In [ ]:
# import os 
# import re 
# import numpy as np 
# import pandas as pd 
# from collections import Counter 
# import torch 
# import torch.nn as nn 
# from torch.utils.data import Dataset, DataLoader 
# from sklearn.model_selection import train_test_split

**config**

In [ ]:
# MAX_LEN = 256 
# EMBED_DIM = 256 
# HIDDEN_DIM = 256 
# BATCH_SIZE = 64 
# EPOCHS = 10 
# LR = 1e-3 
# LABELS = ["A", "B", "C", "D", "E"]

In [ ]:
# #Tokenizer

# def clean_text(text):
#     text = str(text).lower()
#     text = re.sub(r"[^a-z0-9 ]", " ", text)
#     text = re.sub(r"\s+", " ", text).strip()
#     return text

# def tokenize(text):
#     return clean_text(text).split()


In [ ]:
# #Vocab

# def build_vocab(df):

#     counter = Counter()

#     for _, row in df.iterrows():

#         counter.update(tokenize(row["prompt"]))

#         for col in LABELS:
#             counter.update(tokenize(row[col]))

#     vocab = {
#         "<PAD>": 0,
#         "<UNK>": 1
#     }

#     for word, freq in counter.items():
#         if freq >= 2:
#             vocab[word] = len(vocab)

#     return vocab

# def encode(text, vocab):

#     tokens = tokenize(text)

#     ids = [vocab.get(t, 1) for t in tokens]

#     ids = ids[:MAX_LEN]

#     if len(ids) < MAX_LEN:
#         ids += [0] * (MAX_LEN - len(ids))

#     return ids


In [ ]:
# #Dataset

# class MCQDataset(Dataset):

#     def __init__(self, df, vocab, train_mode=True):

#         self.samples = []

#         for _, row in df.iterrows():

#             prompt = str(row["prompt"])

#             if train_mode:

#                 answer = row["answer"]

#                 for label in LABELS:

#                     option = str(row[label])

#                     text = prompt + " [SEP] " + option

#                     target = 1 if label == answer else 0

#                     self.samples.append(
#                         (
#                             encode(text, vocab),
#                             target
#                         )
#                     )

#             else:

#                 question_id = row["id"]

#                 for label in LABELS:

#                     option = str(row[label])

#                     text = prompt + " [SEP] " + option

#                     self.samples.append(
#                         (
#                             question_id,
#                             label,
#                             encode(text, vocab)
#                         )
#                     )

#         self.train_mode = train_mode

#     def __len__(self):
#         return len(self.samples)

#     def __getitem__(self, idx):

#         if self.train_mode:

#             x, y = self.samples[idx]

#             return (
#                 torch.tensor(x, dtype=torch.long),
#                 torch.tensor(y, dtype=torch.float)
#             )

#         qid, label, x = self.samples[idx]

#         return (
#             qid,
#             label,
#             torch.tensor(x, dtype=torch.long)
#         )

In [ ]:
# # Attention

# class AttentionPool(nn.Module):

#     def __init__(self, hidden_size):

#         super().__init__()

#         self.attn = nn.Linear(hidden_size, 1)

#     def forward(self, x):

#         scores = self.attn(x)

#         weights = torch.softmax(scores, dim=1)

#         pooled = (weights * x).sum(dim=1)

#         return pooled

In [ ]:
# # Model

# class MCQRanker(nn.Module):

#     def __init__(self, vocab_size):

#         super().__init__()

#         self.embedding = nn.Embedding(
#             vocab_size,
#             EMBED_DIM,
#             padding_idx=0
#         )

#         self.lstm = nn.LSTM(
#             EMBED_DIM,
#             HIDDEN_DIM,
#             batch_first=True,
#             bidirectional=True
#         )

#         self.pool = AttentionPool(HIDDEN_DIM * 2)

#         self.classifier = nn.Sequential(
#             nn.Linear(HIDDEN_DIM * 2, 256),
#             nn.ReLU(),
#             nn.Dropout(0.3),
#             nn.Linear(256, 1)
#         )

#     def forward(self, x):

#         emb = self.embedding(x)

#         out, _ = self.lstm(emb)

#         pooled = self.pool(out)

#         logits = self.classifier(pooled)

#         return logits.squeeze(-1)

In [ ]:
# #Train...................
# def train_epoch(model, loader, optimizer, criterion):

#     model.train()

#     total_loss = 0

#     for x, y in loader:

#         x = x.to(DEVICE)
#         y = y.to(DEVICE)

#         optimizer.zero_grad()

#         logits = model(x)

#         loss = criterion(logits, y)

#         loss.backward()

#         optimizer.step()

#         total_loss += loss.item()

#     return total_loss / len(loader)

In [ ]:
# #Validation
# @torch.no_grad()
# def evaluate(model, loader, criterion):

#     model.eval()

#     total_loss = 0

#     for x, y in loader:

#         x = x.to(DEVICE)
#         y = y.to(DEVICE)

#         logits = model(x)

#         loss = criterion(logits, y)

#         total_loss += loss.item()

#     return total_loss / len(loader)


In [ ]:
# #Training the model................
# train_df = pd.read_csv("/kaggle/input/competitions/smart-mcq-solver-challenge/train.csv")

# vocab = build_vocab(train_df)

# train_part, valid_part = train_test_split(
#     train_df,
#     test_size=0.1,
#     random_state=42,
#     stratify=train_df["answer"]
# )

# train_ds = MCQDataset(train_part, vocab, True)
# valid_ds = MCQDataset(valid_part, vocab, True)

# train_loader = DataLoader(
#     train_ds,
#     batch_size=BATCH_SIZE,
#     shuffle=True
# )

# valid_loader = DataLoader(
#     valid_ds,
#     batch_size=BATCH_SIZE
# )

# model = MCQRanker(len(vocab)).to(DEVICE)

# criterion = nn.BCEWithLogitsLoss()

# optimizer = torch.optim.Adam(
#     model.parameters(),
#     lr=LR
# )

# best_loss = 999

# for epoch in range(EPOCHS):

#     train_loss = train_epoch(
#         model,
#         train_loader,
#         optimizer,
#         criterion
#     )

#     valid_loss = evaluate(
#         model,
#         valid_loader,
#         criterion
#     )

#     print(
#         f"Epoch {epoch+1}/{EPOCHS} "
#         f"Train_loss={train_loss:.4f} "
#         f"Validation_loss={valid_loss:.4f}"
#     )

#     if valid_loss < best_loss:

#         best_loss = valid_loss

#         torch.save(
#             {
#                 "model": model.state_dict(),
#                 "vocab": vocab
#             },
#             "best_model.pt"
#         )

In [ ]:
# #Inference
# checkpoint = torch.load(
#     "best_model.pt",
#     map_location=DEVICE
# )

# model.load_state_dict(checkpoint["model"])
# model.eval()

# test_df = pd.read_csv('/kaggle/input/competitions/smart-mcq-solver-challenge/test.csv')

# test_ds = MCQDataset(
#     test_df,
#     vocab,
#     train_mode=False
# )

# test_loader = DataLoader(
#     test_ds,
#     batch_size=256,
#     shuffle=False
# )

# # Store scores for each question
# scores = {}

# with torch.no_grad():

#     for qids, labels, x in test_loader:

#         x = x.to(DEVICE)

#         logits = model(x)

#         probs = torch.sigmoid(logits).cpu().numpy()

#         for qid, label, p in zip(qids, labels, probs):

#             # Convert ID to string for consistency
#             qid = str(qid)

#             if qid not in scores:
#                 scores[qid] = {}

#             scores[qid][label] = float(p)
# predictions = []

# for qid in test_df["id"]:

#     qid = str(qid)

#     # Safety fallback
#     if qid not in scores:

#         predictions.append("A B C")
#         continue

#     ranked = sorted(
#         scores[qid].items(),
#         key=lambda x: x[1],
#         reverse=True
#     )

#     top3 = [label for label, score in ranked[:3]]

#     # If somehow fewer than 3 labels exist
#     while len(top3) < 3:
#         for lbl in ["A", "B", "C", "D", "E"]:
#             if lbl not in top3:
#                 top3.append(lbl)
#             if len(top3) == 3:
#                 break

#     predictions.append(" ".join(top3))

**Submission**

In [ ]:
# submission = pd.DataFrame(
#     {
#         "ID": test_df["id"],
#         "Prediction": predictions
#     }
# )

# submission.to_csv(
#     "submission.csv",
#     index=False
# )

# print(submission.head())
# print("submission.csv is created successfully")

## Milestone-1 ##

**Q1**:
Calculate the frequency distribution of the correct answer (A, B, C, D, E) in train.csv. Based on your counts, what is the sum of the occurrences of the most frequent option and the least frequent option?

In [ ]:
# freq = train["answer"].value_counts().sort_index()
# print(freq)



In [ ]:
# result = freq.max() + freq.min()
# print("Sum =", result)

ANS:Sum = 814

**Q2:**
After converting the prompt column to lowercase and removing all standard punctuation characters (using Python's string.punctuation), split the text by whitespace. What is the total number of unique words (vocabulary size) across the entire cleaned prompt column of train.csv?  

In [ ]:
# import string
# vocab = set()

# for text in train["prompt"].astype(str):
#     # lowercase
#     text = text.lower()

#     # remove standard punctuation
#     text = text.translate(str.maketrans('', '', string.punctuation))  #Replace nothing,with nothing,Delete all punctuation characters

#     # split by whitespace
#     vocab.update(text.split())

# print("Vocabulary size:", len(vocab))

ANS:Vocabulary size: 859

**Q3:**
Using the cleaned prompt from Row ID 1, filter out the standard English stop words using sklearn.feature_extraction.text.ENGLISH_STOP_WORDS. How many words are left in the prompt for Row ID 1 after filtering?  

In [ ]:
# from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS

# prompt = train.loc[train["id"] == 1, "prompt"].iloc[0]

# # clean
# prompt = prompt.lower()
# prompt = prompt.translate(str.maketrans('', '', string.punctuation))

# # remove stop words
# words = [word for word in prompt.split() if word not in ENGLISH_STOP_WORDS]

# print(words)
# print(len(words))

ANS:13

**Q4:**
Fit a default TfidfVectorizer(stop_words='english') on a list containing all the combined text of the prompts and options in train.csv. What is the exact total number of feature columns (vocabulary size) generated by the vectorizer? 

In [ ]:
# from sklearn.feature_extraction.text import TfidfVectorizer
# # combine prompt + options
# texts = (
#     train["prompt"] + " " +
#     train["A"] + " " +
#     train["B"] + " " +
#     train["C"] + " " +
#     train["D"] + " " +
#     train["E"]).tolist()

# vect= TfidfVectorizer(stop_words="english")
# X = vect.fit_transform(texts)
# print("Vocabulary size:", len(vect.vocabulary_))
# print("Feature matrix shape:", X.shape)

ANS:Vocabulary size: 2762

**Q5:**
Using the TF-IDF vectorizer fitted in Question 3, calculate the cosine similarity between the prompt and option A strictly for Row ID 1. What is the resulting similarity score? (Round to 4 decimal places).  

In [ ]:
# from sklearn.metrics.pairwise import cosine_similarity
# row = train[train["id"] == 1].iloc[0]

# prompt_text = row["prompt"]
# option_A_text = row["A"]

# v_prompt = vect.transform([prompt_text])
# v_A = vect.transform([option_A_text])
# score = cosine_similarity(v_prompt, v_A)[0][0]

# print(round(score, 5))

ANS:0.27202

**Q6:**
Expand the logic from Question 4: For every row in train.csv, calculate the cosine similarity between the prompt and each of its 5 options .  Then calculate the percentage of instances where the option with the highest cosine similarity matches the correct answer.   

In [ ]:
# vect.fit(texts)
# options = ["A", "B", "C", "D", "E"]
# correct = 0

# for i, row in train.iterrows():
#     prompt_vec = vect.transform([row["prompt"]])

#     similarities = []

#     for opt in options:
#         opt_vec = vect.transform([row[opt]])
#         sim = cosine_similarity(prompt_vec, opt_vec)[0][0]
#         similarities.append(sim)

#     predicted_index = np.argmax(similarities)  #argmax() finds highest value
#     predicted_answer = options[predicted_index]

#     if predicted_answer == row["answer"]:
#         correct += 1
# accuracy = correct / len(train) * 100
# print("Accuracy:", round(accuracy, 2), "%")  #This tells in what % of questions, the option most similar to the prompt was actually correct

ANS:Accuracy: 13.55 %

**Q7:**
If the ground truth answer for a question is C, what is the MAP@3 score if a model predicts C A B?  
**ANS:**
Correct answer C is at rank 1,

so,MAP@3=1/1​

        =1

**Q8:**
If the ground truth answer for a question is  B, what is the MAP@3 score if a model predicts D B E?  

**ANS:**
Rank of B = 2

MAP@3=1/2

=0.5

**Q9:**
The Majority Class Baseline: Find the most frequent correct answer in the training set (using your data from Q1). Make a static prediction for every single row where that most frequent answer is your 1st guess, followed by the second most frequent, and then the third most frequent. What is the overall MAP@3 score of this "Majority Class" baseline on train.csv?

In [ ]:
# freq = train["answer"].value_counts()
# top3 = freq.index[:3].tolist()

# map3 = train["answer"].apply(
#     lambda x: 1 if x==top3[0] else (0.5 if x==top3[1] else (1/3 if x==top3[2] else 0))
# ).mean()

# print(map3)

ANS:0.42125

**Q10:**
The TF-IDF Pipeline: Build a basic pipeline that evaluates every row in train.csv. For each row, calculate the TF-IDF cosine similarity between the prompt and each of the 5 options. Sort these options from highest similarity to lowest to form your top 3 predictions. What is the final average MAP@3 score of this TF-IDF pipeline across the entire training set?  

In [ ]:
# def map3_score(true, preds):  #MAP@3 function
#     if true == preds[0]:
#         return 1.0
#     elif true == preds[1]:
#         return 0.5
#     elif true == preds[2]:
#         return 1/3
#     else:
#         return 0.0

# #Evaluate all rows
# scores = []

# for _,row in train.iterrows():

#     prompt_vec = vect.transform([row["prompt"]])

#     sims = []

#     for opt in options:
#         opt_vec = vect.transform([row[opt]])
#         sim = cosine_similarity(prompt_vec, opt_vec)[0][0]
#         sims.append((opt, sim))

#     # sort by similarity (descending)
#     ranked = sorted(sims, key=lambda x: x[1], reverse=True)

#     top3 = [r[0] for r in ranked[:3]]

#     scores.append(map3_score(row["answer"], top3))

# #Final MAP@3
# final_map3 = np.mean(scores)

# print("TF-IDF Pipeline MAP@3:", round(final_map3, 4))

ANS:TF-IDF Pipeline MAP@3: 0.2962

## Milestone-2 ##

**Q1:**
Load train.csv using the Hugging Face datasets library (do not use pandas). Use the .map() function to create a new column called combined_text that concatenates the prompt and A columns with a space in between. E.g., prompt_text A_text. What is the exact character length (total number of string characters using Python's len() function, NOT the number of tokens) of the combined_text string for the row at index 51? 

In [2]:
# from datasets import load_dataset

# # Load the CSV
# dataset = load_dataset("csv", data_files="/kaggle/input/competitions/smart-mcq-solver-challenge/train.csv")["train"]

# # Create the combined_text column
# dataset = dataset.map(
#     lambda x: {
#         "combined_text": x["prompt"] + " " + x["A"]
#     }
# )

# # Get the character length for row 51
# length = len(dataset[51]["combined_text"])

# print("Length is:",length)

Length is: 614


Ans:Length is: 614

**Q2:**
Initialize the bert-base-uncased tokenizer. Look at the tokenizer's configuration properties: what is the exact total vocabulary size (the maximum number of unique subword tokens the model knows) hardcoded into this tokenizer?  

In [3]:
# from transformers import AutoTokenizer

# tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")

# print(tokenizer.vocab_size)

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

30522


Ans: 30522

Q3:
Transformers rely on special tokens to understand sentence boundaries. Using the bert-base-uncased tokenizer from the previous step, extract the exact integer ID assigned to the [SEP] (Separator) token.  

In [4]:
# print(tokenizer.sep_token)     
# print(tokenizer.sep_token_id)

[SEP]
102


Ans: 102

Q4:
Using the bert-base-uncased tokenizer, tokenize the entire prompt column of the train dataset simultaneously. Set padding='max_length', truncation=True, max_length=128, and return_tensors='pt' (PyTorch tensors). 

What is the exact geometric shape (dimensions) of the resulting input_ids tensor?

In [14]:
# from transformers import AutoTokenizer

# tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")

# prompts = [
#     "" if x is None else str(x)
#     for x in train_dataset["prompt"]
# ]

# encodings = tokenizer(
#     prompts,
#     padding="max_length",
#     truncation=True,
#     max_length=128,
#     return_tensors="pt"
# )

# print(encodings["input_ids"].shape)

torch.Size([2000, 128])


Ans: torch.Size([2000, 128])

Q5:
A standard bert-base-uncased model has a hidden embedding size of 768 dimensions and uses exactly 12 attention heads in each layer. 

In Transformer architecture, the hidden size is divided equally among the attention heads. What is the exact dimensionality (size) of each individual attention head?  

In [16]:
# from transformers import AutoConfig

# config = AutoConfig.from_pretrained("bert-base-uncased")

# print(config.hidden_size)          # 768
# print(config.num_attention_heads)  # 12
# print("Ans:",config.hidden_size // config.num_attention_heads) 

768
12
Ans: 64


Ans: 64

Q6:
Load the bert-base-uncased model using AutoModel.from_pretrained(). Tokenize the prompt from row ID 0 using the tokenizer's default settings (do not apply any manual padding or truncation). Pass this tokenized input through the model. Look at the output object. 

What is the exact shape of the last_hidden_state tensor returned? 

Note: We follow zero-indexing here.

In [17]:
# from datasets import load_dataset
# from transformers import AutoTokenizer, AutoModel

# # Load dataset
# train_dataset = load_dataset(
#     "csv",
#     data_files="/kaggle/input/competitions/smart-mcq-solver-challenge/train.csv"
# )["train"]

# # Load tokenizer and model
# tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")
# model = AutoModel.from_pretrained("bert-base-uncased")

# # Tokenize prompt at row 0 (default settings)
# inputs = tokenizer(train_dataset[0]["prompt"], return_tensors="pt")

# # Forward pass
# outputs = model(**inputs)

# print(inputs["input_ids"].shape)
# print(outputs.last_hidden_state.shape)

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

BertModel LOAD REPORT from: bert-base-uncased
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
cls.predictions.bias                       | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED |  | 
cls.seq_relationship.bias                  | UNEXPECTED |  | 
cls.seq_relationship.weight                | UNEXPECTED |  | 
cls.predictions.transform.dense.weight     | UNEXPECTED |  | 
cls.predictions.transform.dense.bias       | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


torch.Size([1, 31])
torch.Size([1, 31, 768])


In [18]:
# print(outputs.last_hidden_state.shape)

torch.Size([1, 31, 768])


Ans: torch.Size([1, 31, 768])

**Q7:**
Using the last_hidden_state tensor from the previous question, extract the embedding vector representing the [CLS] token (which is always the token at index 0). What is the sum of the first 5 float values in this [CLS] vector? (Round your answer to 4 decimal places).  

In [19]:
# from datasets import load_dataset
# from transformers import AutoTokenizer, AutoModel
# import torch

# # Load dataset
# train_dataset = load_dataset(
#     "csv",
#     data_files="/kaggle/input/competitions/smart-mcq-solver-challenge/train.csv"
# )["train"]

# # Load tokenizer and model
# tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")
# model = AutoModel.from_pretrained("bert-base-uncased")

# # Put model in evaluation mode
# model.eval()

# # Tokenize row 0 prompt (default tokenizer settings)
# inputs = tokenizer(train_dataset[0]["prompt"], return_tensors="pt")

# # Forward pass
# with torch.no_grad():
#     outputs = model(**inputs)

# # Extract the [CLS] embedding
# cls_embedding = outputs.last_hidden_state[0, 0]

# # Sum the first 5 values
# answer = cls_embedding[:5].sum().item()

# print("First 5 values:", cls_embedding[:5])
# print("Sum:", round(answer, 4))

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

BertModel LOAD REPORT from: bert-base-uncased
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
cls.predictions.bias                       | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED |  | 
cls.seq_relationship.bias                  | UNEXPECTED |  | 
cls.seq_relationship.weight                | UNEXPECTED |  | 
cls.predictions.transform.dense.weight     | UNEXPECTED |  | 
cls.predictions.transform.dense.bias       | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


First 5 values: tensor([-0.4677, -0.0754, -0.2019, -0.0071, -0.4480])
Sum: -1.2001


Ans: Sum= -1.2001

Q8:
Load bert-base-uncased with the parameter output_attentions=True. Tokenize the exact string "Light-ion fusion is a technique." (ensuring you set return_tensors='pt') and pass it through the model. Extract the attention matrix for the last layer (index -1) and the first attention head (head index 0). 

What is the exact attention weight (a float value) that the [CLS] token (token index 0) pays to the word fusion (you will need to find the specific token index for fusion in the input_ids)? (Round your answer to 4 decimal places).  

In [20]:
# from transformers import AutoTokenizer, AutoModel
# import torch

# # Load tokenizer and model
# tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")
# model = AutoModel.from_pretrained(
#     "bert-base-uncased",
#     output_attentions=True
# )

# model.eval()

# text = "Light-ion fusion is a technique."

# # Tokenize
# inputs = tokenizer(text, return_tensors="pt")

# # Show tokens and their indices
# tokens = tokenizer.convert_ids_to_tokens(inputs["input_ids"][0])
# for i, token in enumerate(tokens):
#     print(i, token)

# # Forward pass
# with torch.no_grad():
#     outputs = model(**inputs)

# # Last layer, first head
# attention = outputs.attentions[-1][0, 0]

# # Find the index of "fusion"
# fusion_index = tokens.index("fusion")

# # Attention from [CLS] (index 0) to "fusion"
# weight = attention[0, fusion_index].item()

# print("Attention weight:", round(weight, 4))

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

BertModel LOAD REPORT from: bert-base-uncased
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
cls.predictions.bias                       | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED |  | 
cls.seq_relationship.bias                  | UNEXPECTED |  | 
cls.seq_relationship.weight                | UNEXPECTED |  | 
cls.predictions.transform.dense.weight     | UNEXPECTED |  | 
cls.predictions.transform.dense.bias       | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


0 [CLS]
1 light
2 -
3 ion
4 fusion
5 is
6 a
7 technique
8 .
9 [SEP]
Attention weight: 0.1025


Ans: Attention weight: 0.1025

Q9:
Initialize the sentence-transformers/all-MiniLM-L6-v2 model. Use the model's .encode() method to generate embeddings for both the prompt and Option B for row ID 0. Calculate the cosine similarity between these two vectors specifically using the sentence_transformers.util.cos_sim() function. What is the resulting similarity score rounded to 4 decimal places? Note: We follow zero-indexing here.

In [21]:
# from datasets import load_dataset
# from sentence_transformers import SentenceTransformer, util

# # Load dataset
# train_dataset = load_dataset(
#     "csv",
#     data_files="/kaggle/input/competitions/smart-mcq-solver-challenge/train.csv"
# )["train"]

# # Load model
# model = SentenceTransformer("sentence-transformers/all-MiniLM-L6-v2")

# # Get row 0
# prompt = train_dataset[0]["prompt"]
# option_b = train_dataset[0]["B"]

# # Generate embeddings
# prompt_emb = model.encode(prompt, convert_to_tensor=True)
# option_b_emb = model.encode(option_b, convert_to_tensor=True)

# # Cosine similarity
# score = util.cos_sim(prompt_emb, option_b_emb)

# print("Similarity:", round(score.item(), 4))

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

Similarity: 0.7658


Ans: Similarity: 0.7658

Q10:
Build two complete ranking pipelines evaluating every row in train.csv.

Pipeline 1: Use the TF-IDF cosine similarity approach from Milestone 1.

Pipeline 2: Use the sentence-transformers/all-MiniLM-L6-v2 model to generate embeddings for the prompt and all five options. Rank options using cosine similarity to form Top-3 predictions.

First, what is the final MAP@3 score of the all-MiniLM-L6-v2 pipeline across the entire training set? 

Second, count the number of questions for which the correct answer is NOT present in the TF-IDF Top-3 predictions BUT IS present in the MiniLM Top-3 predictions. What is this exact resulting count? 

In [22]:
# from datasets import load_dataset
# from sklearn.feature_extraction.text import TfidfVectorizer
# from sklearn.metrics.pairwise import cosine_similarity
# from sentence_transformers import SentenceTransformer, util
# import torch

# # Load dataset
# train = load_dataset(
#     "csv",
#     data_files="/kaggle/input/competitions/smart-mcq-solver-challenge/train.csv"
# )["train"]

# letters = ["A", "B", "C", "D", "E"]

# #-------------------------------------------------------------------------
# # MAP@3
# def map3(predictions, truths):
#     score = 0.0
#     for pred, truth in zip(predictions, truths):
#         if truth in pred[:3]:
#             score += 1.0 / (pred.index(truth) + 1)
#     return score / len(truths)

# #-----------------------------------------------------------
# # Pipeline 1 :TF-IDF

# tfidf_preds = []

# for row in train:
#     docs = [row["prompt"]] + [row[c] for c in letters]

#     vec = TfidfVectorizer().fit_transform(docs)

#     sims = cosine_similarity(vec[0:1], vec[1:]).flatten()

#     ranking = [letters[i] for i in sims.argsort()[::-1]]

#     tfidf_preds.append(ranking)

# #----------------------------------------------------------------
# # Pipeline 2 :MiniLM

# model = SentenceTransformer("sentence-transformers/all-MiniLM-L6-v2")

# minilm_preds = []

# for row in train:

#     prompt_emb = model.encode(
#         row["prompt"],
#         convert_to_tensor=True,
#         show_progress_bar=False
#     )

#     option_embs = model.encode(
#         [row[c] for c in letters],
#         convert_to_tensor=True,
#         show_progress_bar=False
#     )

#     sims = util.cos_sim(prompt_emb, option_embs)[0]

#     ranking = [
#         letters[i]
#         for i in torch.argsort(sims, descending=True).tolist()
#     ]

#     minilm_preds.append(ranking)

# #-------------------------------------------
# # Results
# answers = train["answer"]

# map_minilm = map3(minilm_preds, answers)

# improved = 0

# for tfidf_rank, mini_rank, ans in zip(tfidf_preds, minilm_preds, answers):

#     if ans not in tfidf_rank[:3] and ans in mini_rank[:3]:
#         improved += 1

# print("MiniLM MAP@3 =", round(map_minilm, 6))
# print("Improved Count =", improved)

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


MiniLM MAP@3 = 0.423083
Improved Count = 564


Ans: Improved Count = 564

Q11:
Initialize the Hugging Face pipeline for "zero-shot-classification" (it will default to facebook/bart-large-mnli). For the prompt of the 2nd row (index 1), pass Options A, B, and C as the candidate_labels. What is the probability score given to the top-ranked option? (Round to 4 decimal places).

In [23]:
# from datasets import load_dataset
# from transformers import pipeline

# # Load dataset
# train = load_dataset(
#     "csv",
#     data_files="/kaggle/input/competitions/smart-mcq-solver-challenge/train.csv"
# )["train"]

# # Initialize zero-shot classifier
# classifier = pipeline("zero-shot-classification")

# # Row index 1
# sequence = train[1]["prompt"]

# candidate_labels = [
#     train[1]["A"],
#     train[1]["B"],
#     train[1]["C"]
# ]

# # Run inference
# result = classifier(sequence, candidate_labels)

# print(result["labels"])
# print(result["scores"])

# # Top probability
# print(round(result["scores"][0], 4))

No model was supplied, defaulted to facebook/bart-large-mnli and revision d7645e1.
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/1.63G [00:00<?, ?B/s]

Loading weights:   0%|          | 0/515 [00:00<?, ?it/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

['Accelerator-based light-ion fusion is a technique that uses particle accelerators to achieve particle kinetic energies sufficient to induce light-ion fusion reactions. This method is relatively easy to implement and can be done in an efficient manner, requiring only a vacuum tube, a pair of electrodes, and a high-voltage transformer. Fusion can be observed with as little as 10 kV between the electrodes.', 'Accelerator-based light-ion fusion is a technique that uses particle accelerators to achieve particle kinetic energies sufficient to induce light-ion fusion reactions. This method is relatively difficult to implement and requires a complex system of vacuum tubes, electrodes, and transformers. Fusion can be observed with as little as 100 kV between the electrodes.', 'Accelerator-based light-ion fusion is a technique that uses particle accelerators to achieve particle kinetic energies sufficient to induce heavy-ion fusion reactions. This method is relatively difficult to implement an

Ans: 0.4575

Q12:
Run the exact same zero-shot classification as the previous question, but this time pass the argument multi_label=True. 

What is the absolute difference between the sum of the 3 probabilities in the previous question (which uses Softmax) and the sum of the 3 probabilities in this question (which uses independent Sigmoids)?

In [24]:
# from datasets import load_dataset
# from transformers import pipeline

# # Load dataset
# train = load_dataset(
#     "csv",
#     data_files="/kaggle/input/competitions/smart-mcq-solver-challenge/train.csv"
# )["train"]

# # Zero-shot pipeline (facebook/bart-large-mnli)
# classifier = pipeline("zero-shot-classification")

# # Row index 1
# sequence = train[1]["prompt"]
# candidate_labels = [
#     train[1]["A"],
#     train[1]["B"],
#     train[1]["C"]
# ]

# # Previous question: Softmax (default)
# softmax_result = classifier(
#     sequence,
#     candidate_labels
# )

# softmax_sum = sum(softmax_result["scores"])

# # This question: Independent Sigmoids
# sigmoid_result = classifier(
#     sequence,
#     candidate_labels,
#     multi_label=True
# )

# sigmoid_sum = sum(sigmoid_result["scores"])

# difference = abs(softmax_sum - sigmoid_sum)

# print("Softmax scores:", softmax_result["scores"])
# print("Softmax sum:", softmax_sum)

# print("Sigmoid scores:", sigmoid_result["scores"])
# print("Sigmoid sum:", sigmoid_sum)

# print("Absolute difference:", round(difference, 4))

No model was supplied, defaulted to facebook/bart-large-mnli and revision d7645e1.
Using a pipeline without specifying a model name and revision in production is not recommended.


Loading weights:   0%|          | 0/515 [00:00<?, ?it/s]

Softmax scores: [0.4574522376060486, 0.2750644385814667, 0.26748329401016235]
Softmax sum: 0.9999999701976776
Sigmoid scores: [0.00046926175127737224, 2.063553074549418e-05, 1.9700279153767042e-05]
Sigmoid sum: 0.0005095975611766335
Absolute difference: 0.9995


Ans: Absolute difference: 0.9995

**Q13:**
Let's try Generative AI instead of Classification. 

Load a Small Language Model like google/flan-t5-small using the Hugging Face pipeline("text2text-generation"). Construct the following exact string for row index 0: "Question: [prompt]. Is the correct answer A: [A] or B: [B]? Answer with just the letter A or B." 
Pass this string to the pipeline, setting max_new_tokens=5. What is the exact string output returned by the model? 

In [30]:
# from datasets import load_dataset
# from transformers import AutoTokenizer, AutoModelForSeq2SeqLM

# # Load dataset
# train = load_dataset(
#     "csv",
#     data_files="/kaggle/input/competitions/smart-mcq-solver-challenge/train.csv"
# )["train"]

# # Load model
# tokenizer = AutoTokenizer.from_pretrained("google/flan-t5-small")
# model = AutoModelForSeq2SeqLM.from_pretrained("google/flan-t5-small")

# # Row 0
# row = train[0]

# prompt = (
#     f"Question: {row['prompt']}. "
#     f"Is the correct answer A: {row['A']} "
#     f"or B: {row['B']}? "
#     f"Answer with just the letter A or B."
# )

# # Tokenize
# inputs = tokenizer(prompt, return_tensors="pt")

# # Generate
# outputs = model.generate(
#     **inputs,
#     max_new_tokens=5
# )

# # Decode
# answer = tokenizer.decode(outputs[0], skip_special_tokens=True)

# print("Ans:",answer)

Loading weights:   0%|          | 0/190 [00:00<?, ?it/s]

The tied weights mapping and config for this model specifies to tie shared.weight to lm_head.weight, but both are present in the checkpoints, so we will NOT tie them. You should update the config with `tie_word_embeddings=False` to silence this warning


Ans: B


Ans: B